### PR TITLE
Allow `register` to be called after `onload` event

### PR DESCRIPTION
**Pull request by cainrus**
It's not possible to register service-worker on demand right now.

### CODE REVIEW

**Comment by sodatea** [comment]
I'm sorry but I don't understand what difference does it make?

**Comment by cainrus** [comment]
> I'm sorry but I don't understand what difference does it make?

if you call register function on click event, then service-worker will not register itself, because `load` event has been triggered  already. Bug has been fixed 🤓

**Comment by kmohrf** [comment]
This is helpful if the service worker registration is an optional feature in a web app. In my case we produce a single build that can be configured at runtime through a JSON file. If the organization that uses the software provided the necessary configuration options the service worker is registered and the web app manifest is injected into the page. But as the configuration is loaded after the load event has been triggered, register-service-worker will never actually register the service worker.

**Comment by eduardosada** [comment]
Would be nice to have this PR merged.

**Comment by nicolas-t** [comment]
I expected the registration to happen when I called the `register` function.
Not after an "arbitrary" event of the page lifecycle.

The proposed solution is nice because it's not breaking, and should still be compatible with vue-cli PWA plugin generated file `registerServiceWorker.js`
 
On the other hand, I think it'd be more flexible (and intuitive) to let the parent script handle the timing of the registration.

Something like this: 
```js
import { register } from 'register-service-worker'

window.addEventListener('load', () => {
    register('/service-worker.js', { ... })
})
```

Which would also allow this (for example) : 

```vue
<script>
import { register } from 'register-service-worker'

export default {
    // ...
    
    mounted() {
        register('/service-worker.js', { ... })
    }
}
</script>
```

**Comment by cainrus** [comment]
@nicolas-t 
> On the other hand, I think it'd be more flexible (and intuitive) to let the parent script handle the timing of the registration.

I believe it's a breaking change for next major version.

**Comment by sodatea** [comment]
Thanks for the help. I've also added a compatibility layer to avoid `ReferenceError` in IE11.
This feature will be landed in the 1.7 release.

File: src/index.js [hunk 1]
```diff
@@ -15,6 +15,17 @@ const isLocalhost = () => Boolean(
...
+let waitWindowLoad
+// Typically, a browser that supports `serviceWorker` should also have supported
+// `Promise`. But as this package can be used in environments without service
+// worker support (in that case it would do nothing), there's a chance that
+// `Promise` does not exist. So we must check for its existence first.
+if (typeof Promise !== 'undefined') {
+  waitWindowLoad = new Promise(resolve => window.addEventListener('load', resolve))
+} else {
+  waitWindowLoad = { then: (cb) => window.addEventListener('load', cb) }
+}
```
Comment (meteorlxy) [inline src/index.js]: These code reference `window` in the outter scope, which cause potential breaking changes.

For example, it breaks the SSR process in VuePress https://github.com/vuejs/vuepress/issues/2222

@sodatea

Comment (sodatea) [reply src/index.js]: Thanks. I'll fix it soon.

**Pull request merged**